### PR TITLE
[Model] SiglipVisionModel ported from transformers

### LINKED ISSUE #6941: [Feature]: SiglipVisionModel Support
### 🚀 The feature, motivation and pitch

Paligemma uses SiglipVisionModel from the transformers library, which can be ported to the vLLM.
As CLIPVisionModel is supported for VLMs, the SiglipVisionModel can be supported too.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ChristopherCho**
This PR implemented SiglipVisionModel for VLMs.
- Some of the pre-trained SiglipVisionModel cannot use vLLM's Attention layer.
    Therefore, I implemented alternative attention layers if vLLM's one is impossible.
- I tried vllm_flash_attn backend which doesn't work properly with CUDA Error.
    Thus, only the basic attention mechanism is working properly for now.
- Modified Paligemma to use implemented SiglipVisionModel.

FIX #6941
FIX #7144

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Just saw this. Thanks for the implemention, I'll leave the review to @ywang96 since he worked on PaliGemma.

**Comment by jeejeelee** [comment]
The calculation of attention can use the MEA (Memory Efficient Attention) ops  from xformers.  see: https://facebookresearch.github.io/xformers/components/ops.html

**Comment by ChristopherCho** [comment]
@jeejeelee 
> The calculation of attention can use the MEA (Memory Efficient Attention) ops from xformers. see: https://facebookresearch.github.io/xformers/components/ops.html

Thanks! I added xformers MEA and torch sdpa to give various options.

**Comment by jeejeelee** [comment]
Hi, thank you for your excellent work. I'd like to know if your implementation supports the following model:
```python
import timm 
 model = timm.create_model(
                "vit_so400m_patch14_siglip_384.webli",
                pretrained=False,
                num_classes=0,
                dynamic_img_size=True,
                dynamic_img_pad=True,
            )
```

**Comment by ChristopherCho** [comment]
@jeejeelee 
Hi, I believe that the pre-trained Siglip model `vit_so400m_patch14_siglip_384.webli` is just the same as [the one on the Huggingface](https://huggingface.co/google/siglip-so400m-patch14-384).
Also, afaik, the pre-trained Paligemma model uses the same pre-trained Siglip vision encoder.

When I tried loading the pre-trained Paligemma model with the following codes, I could successfully load and infer.
```python
import requests
from PIL import Image
from vllm import LLM, SamplingParams

model_id = "google/paligemma-3b-mix-224"
prompt = "What is on the flower?"
image_file = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/bee.jpg?download=true"
image = Image.open(requests.get(image_file, stream=True).raw)

llm = LLM(model=model_id)
sampling_params = SamplingParams(
    temperature=0.0
)

input_dict = {
    "prompt": prompt,
    "multi_modal_data": {
        "image": image,
    }
}
outputs = llm.generate(input_dict, sampling_params)

print(outputs[0].outputs[0].text)
```

So, as the pre-trained Paligemma works fine, the pre-trained Siglip vision model should work.
But there are some caveats.
1. As the current version of the code is only for supporting VLMs, it does not contain any textual part of the Siglip model (e.g. SiglipTextTransformer, SiglipTextModel, etc.). Therefore, the full Siglip model cannot be loaded with this code.
2. The aforementioned Siglip model is the exact case in which you cannot use vLLM's attention. That is because the head size is 72 (the hidden size of the model is 1152 and the number of heads is 16), which is not in [the supported head sizes](https://github.com/vllm-project/vllm/blob/3c10591ef2e78fbb6aa341195c4b24c36ae8b84d/vllm/attention/ops/paged_attn.py#L36). 
    As I implemented the fallback logic for this case, it works fine but does not use the vLLM's paged attention.

**Comment by ywang96** [comment]
> @jeejeelee Hi, I believe that the pre-trained Siglip model `vit_so400m_patch14_siglip_384.webli` is just the same as [the one on the Huggingface](https://huggingface.co/google/siglip-so400m-patch14-384). Also, afaik, the pre-trained Paligemma model uses the same pre-trained Siglip vision encoder.
> 
> When I tried loading the pre-trained Paligemma model with the following codes, I could successfully load and infer.
> 
> ```python
> import requests
> from PIL import Image
> from vllm import LLM, SamplingParams
> 
> model_id = "google/paligemma-3b-mix-224"
> prompt = "What is on the flower?"
> image_file = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/bee.jpg?download=true"
> image = Image.open(requests.get(image_file, stream=True).raw)
> 
> llm = LLM(model=model_id)
> sampling_params = SamplingParams(
>     temperature=0.0
> )
> 
> input_dict = {
>     "prompt": prompt,
>     "multi_modal_data": {
>         "image": image,
>     }
> }
> outputs = llm.generate(input_dict, sampling_params)
> 
> print(outputs[0].outputs[0].text)
> ```
> 
> So, as the pre-trained Paligemma works fine, the pre-trained Siglip vision model should work. But there are some caveats.
> 
> 1. As the current version of the code is only for supporting VLMs, it does not contain any textual part of the Siglip model (e.g. SiglipTextTransformer, SiglipTextModel, etc.). Therefore, the full Siglip model cannot be loaded with this code.
> 2. The aforementioned Siglip model is the exact case in which you cannot use vLLM's attention. That is because the head size is 72 (the hidden size of the model is 1152 and the number of heads is 16), which is not in [the supported head sizes](https://github.com/vllm-project/vllm/blob/3c10591ef2e78fbb6aa341195c4b24c36ae8b84d/vllm/attention/ops/paged_attn.py#L36).
>    As I implemented the fallback logic for this case, it works fine but does not use the vLLM's paged attention.

Hey @ChristopherCho! Thank you very much for this PR and I really appreciate it - will review this tonight!

File: vllm/model_executor/models/siglip.py [hunk 1]
```diff
@@ -0,0 +1,780 @@
...
+        )
+
+        try:
+            self.attn = Attention(
+                self.num_heads,
+                self.head_dim,
+                self.scale,
+                cache_config=cache_config,
+                quant_config=quant_config,
+            )
```
Comment (ywang96) [inline vllm/model_executor/models/siglip.py:268]: Currently for `ClipVisionModel`, we don't use the vLLM internal `Attention` since the ViT encoder only runs once at the prefill time per sequence, thus I don't think there's much value leveraging a KV cache for this.

Have you seen a significant performance speedup using vLLM `Attention` compared to `transformers` Attention? If not, I think we'd rather just use the one from `transformers` for simplicity for now since this is not the major bottleneck in the whole inference pipeline.

File: vllm/model_executor/models/paligemma.py [hunk 2]
```diff
@@ -308,9 +302,23 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
             use_default_weight_loading = False
             if "vision" in name:
                 if self.vision_tower is not None:
-                    # We only do sharding for language model and
-                    # not vision model for now.
-                    use_default_weight_loading = True
```
Comment (ywang96) [inline vllm/model_executor/models/paligemma.py:313]: Please revert this for now - if we're going to apply TP on the vision tower, we should do it in another separate PR with `CLIPVisionModel` together. Ideally, we should not apply a infrastructure change to only one model.

File: vllm/model_executor/models/siglip.py [hunk 3]
```diff
@@ -0,0 +1,780 @@
...
+
+        return attn_output
+
+
+SIGLIP_ATTENTION_CLASSES = {
+    "eager": SiglipAttention,
+    "flash_attention_2": SiglipFlashAttention2,
+    "sdpa": SiglipSdpaAttention,
+    "xformers": SiglipxFormersAttention,
+}
```
Comment (ywang96) [inline vllm/model_executor/models/siglip.py:446]: I really appreciate that you went out and implemented these (regardless if we're going to use them or not)!

File: vllm/model_executor/models/siglip.py [hunk 4]
```diff
@@ -0,0 +1,780 @@
...
+            hidden_states (`torch.FloatTensor`):
+                Input to the layer of shape `(batch, seq_len, embed_dim)`.
+            attention_mask (`torch.FloatTensor`):
+                Attention mask of shape `(batch, 1, q_len, k_v_seq_len)`
+                where padding elements are indicated by very large negative
+                values.
+            output_attentions (`bool`, *optional*, defaults to `False`):
+                Whether or not to return the attentions tensors of all attention
+                layers. See `attentions` under returned tensors for more detail.
+        """
```
Comment (ywang96) [inline vllm/model_executor/models/siglip.py:523]:
```suggestion
```
Please delete these - I believe you copied the function signature from transformers but forgot to remove them.

File: vllm/model_executor/models/siglip.py [hunk 5]
```diff
@@ -0,0 +1,780 @@
...
+        self,
+        pixel_values,
+        kv_caches: List[torch.Tensor],
+        attn_metadata: AttentionMetadata,
+        interpolate_pos_encoding: Optional[bool] = True,
+    ) -> Tuple:
+        r"""
+        Returns:
+
+        """
```
Comment (ywang96) [inline vllm/model_executor/models/siglip.py:619]:
```suggestion
```

**Review by ywang96 — COMMENTED** [review]
Hey @ChristopherCho - Thank you very much for the PR!

I took a first pass and left some comments. Mostly I'm wondering if we should really use vLLM's attention module in the ViT when it's only used once per sequence, and I suggest simply using the [attention modules](https://github.com/huggingface/transformers/blob/2af199c42b545f6248475ce456dd6c2a351b8522/src/transformers/models/siglip/modeling_siglip.py#L590) from `transformers` for now.

**Comment by ywang96** [comment]
@ChristopherCho Now that we merged #7020 - I think there's some benefit of enabling TP for this model, given that SigLip is 400M parameters in 3B PaliGemma model.

However, could you take a look at the implementation of `attention` [here](https://github.com/vllm-project/vllm/blob/f80ab3521ca2aa74e121e26a27b87da7a1065939/vllm/model_executor/models/idefics2_vision_model.py#L105) instead of using the vLLM `attention`? The latter creates kv cache that I don't think we should be using for VLM when it's only run at the prefill stage.

File: vllm/model_executor/models/siglip.py (hunk 1 above)
Comment (ChristopherCho) [reply vllm/model_executor/models/siglip.py]: Indeed, there weren't significant improvements by using vLLM Attention.
I believe it is due to the reason that you mentioned. It does not leverage the advantages of using a KV cache.
I removed the vLLM Attention part, but keep the log at bb570c3 for the future.

File: vllm/model_executor/models/paligemma.py (hunk 2 above)
Comment (ChristopherCho) [reply vllm/model_executor/models/paligemma.py]: Reverted via dee55d0

**Comment by ChristopherCho** [comment]
@ywang96 I've removed vLLM `Attention` for now and utilized the basic attention mechanism, which is just the same as the [`SiglipAttention`](https://github.com/huggingface/transformers/blob/2af199c42b545f6248475ce456dd6c2a351b8522/src/transformers/models/siglip/modeling_siglip.py#L377-L427).

By the way, I've found that you mentioned the xformers MEA which is implemented [here](https://github.com/ChristopherCho/vllm/blob/dee55d0d9f7c042ad055a731fc6817bcb64dc5eb/vllm/model_executor/models/siglip.py#L400-L420).
Would it be better to utilize this as the baseline attention mechanism? I've tested both the basic attention mechanism and the MEA and found that the MEA was a little bit slower in my environment.

**Comment by ywang96** [comment]
> @ywang96 I've removed vLLM `Attention` for now and utilized the basic attention mechanism, which is just the same as the [`SiglipAttention`](https://github.com/huggingface/transformers/blob/2af199c42b545f6248475ce456dd6c2a351b8522/src/transformers/models/siglip/modeling_siglip.py#L377-L427).
> 
> By the way, I've found that you mentioned the xformers MEA which is implemented [here](https://github.com/ChristopherCho/vllm/blob/dee55d0d9f7c042ad055a731fc6817bcb64dc5eb/vllm/model_executor/models/siglip.py#L400-L420). Would it be better to utilize this as the baseline attention mechanism? I've tested both the basic attention mechanism and the MEA and found that the MEA was a little bit slower in my environment.

Let's use default MHA implementation for this PR: I think if you use MEA then we need to necessarily TP the attention block (since it's using the vLLM TP layers). We can leave a TODO here and do the TP in a later PR!

**Comment by ChristopherCho** [comment]
@ywang96 

Okay, I've implemented the code to use transformers `SiglipAttention` for now, but keep TP versions for TODO in this PR.
By the way, thank you for your comments!

**Comment by ywang96** [comment]
Overall LGTM! I will just need to test this PR locally myself to make sure everything works fine!

**Review by ywang96 — APPROVED** [review]
@ChristopherCho I've made a few more changes to this PR afterwards and verified it works with both TP=1 and TP>1. Thank you again for working on this!
